ocaml/xapi/xapi_network_sriov_helpers.ml

```diff
@@ -124,3 +125,103 @@ let assert_sriov_pif_compatible_with_network ~__context ~pif ~network =
+  let pci = Db.PIF.get_PCI ~__context ~self:physical_PIF in
+  Xapi_pci.get_idle_vf_nums ~__context ~self:pci
+  |> Int64.of_int
+
+let rec get_underlying_pif ~__context ~pif =
+  match Db.PIF.get_sriov_logical_PIF_of ~__context ~self:pif with
+  | [] ->
+    let vlan = Db.PIF.get_VLAN_master_of ~__context ~self:pif in
+    if vlan = Ref.null then None
+    else begin
+      let tagged_pif = Db.VLAN.get_tagged_PIF ~__context ~self:vlan in
+      get_underlying_pif ~__context ~pif:tagged_pif
+    end
+  | sriov :: _ ->
+    Some (Db.Network_sriov.get_physical_PIF ~__context ~self:sriov)
+
+let is_sriov_network ~__context ~self =
+  let pifs = Db.Network.get_PIFs ~__context ~self in
+  match Listext.List.filter_map (fun pif -> get_underlying_pif ~__context ~pif) pifs with
+  | [] -> false
+  | _ -> true
+
+let get_sriov_networks_from_vm ~__context ~vm =
+  let networks = Db.VM.get_VIFs ~__context ~self:vm |> List.map (fun vif -> Db.VIF.get_network ~__context ~self:vif ) in
+  List.filter (fun network -> is_sriov_network ~__context ~self:network) networks
+
+let get_pcis_from_network ~__context ~network =
+  Db.Network.get_PIFs ~__context ~self:network
+  |> Listext.List.filter_map (fun pif -> get_underlying_pif ~__context ~pif)
+  |> List.map (fun pif -> Db.PIF.get_PCI ~__context ~self:pif)
+
+let get_remaining_on_network ~__context ~network =
+  let open Xapi_stdext_monadic in
+  try
+    let pcis = get_pcis_from_network ~__context ~network in
+    let capacity = List.fold_left
+        (fun acc pci ->
+            let remaining_capacity = Xapi_pci.get_idle_vf_nums ~__context ~self:pci |> Int64.of_int in
+            Int64.add acc remaining_capacity
+        ) 0L pcis
+    in
+    if capacity > 0L
+    then Either.Right capacity
+    else Either.Left (Api_errors.Server_error(Api_errors.network_sriov_insufficient_vfs, []))
+  with e -> Either.Left e
+
+let host_sriov_capable ~__context ~host ~network =
+  (*at most one pif in network on localhost*)
+  let local_pifs = Xapi_network_attach_helpers.get_local_pifs ~__context ~network ~host in
+  match local_pifs with
+  | local_pif :: _ ->
+    begin match get_underlying_pif ~__context ~pif:local_pif with
+      | Some underlying_pif ->
+        let pci = Db.PIF.get_PCI ~__context ~self:underlying_pif in
+        if Xapi_pci.get_idle_vf_nums ~__context ~self:pci > 0 then true else false
+      | None -> false
+    end
+  | [] -> false
+
+let group_hosts_by_best_sriov ~__context ~network =
+  let pcis = get_pcis_from_network __context network in
+  let can_allocate_vf pci = Xapi_pci.get_idle_vf_nums ~__context ~self:pci |> Int64.of_int > 0L in
+  let valid_pcis = List.filter can_allocate_vf pcis in
+  (* group hosts by breadth-first algorithm *)
+  Helpers.group_by `descending
+    (fun host ->
+        let pci = List.find (fun pci -> host = Db.PCI.get_host ~__context ~self:pci) valid_pcis in
+        Xapi_pci.get_idle_vf_nums ~__context ~self:pci |> Int64.of_int
```

Comment:
This is the second time we call `PCI.get_host` and `get_idle_vf_nums`. Can't we just skip the `valid_pcis` computation. I'm sure this can be made more efficient.